### DISCUSSION TITLE
[Proposal] A negative case for ?? (which could be use as short cut for ? : operator)

### DISCUSSION THREAD

**Discussion by AliRafiee** [dotnet/csharplang#6330] [Language Ideas] [answered]
The ?? is an awesome operator. I would really like to see the negative of that. Maybe ~?

?? returns the left if it is not null.
~? would return null if the left is null, otherwise it will return the right. The return type is the type of the right parameter and has nothing to do with the type of the left parameter.

This could replace code like this:
```
ClassA possibleNullValue = ....
...
ClassB value = possibleNullValue == null ? null : new ClassB(possibleNullValue);
```

To this:
```
ClassA possibleNullValue = ....
...
ClassB value = possibleNullValue ~? new ClassB(possibleNullValue);
```

(Alternative would be to allow constructors to return null.)

**Comment by TahirAhmadov** [comment]
This is actually a great idea. In the meantime, I have extension methods for this:
```cs
public static TResult? IfNotNull<TObject, TResult>(this TObject? This, Func<TObject, TResult> getResult) where TObject : class where TResult : class
{
  if(This != null) return getResult(This); else return null;
}
// repeat for other combinations of class and struct
string? s = ...;
string? s2 = s.IfNotNull(x=> "prefix" + x[2..4]);
// instead of
string? s2 = s != null ? "prefix" + s[2..4] : null;
// with your proposal
string? s2 = s ~? "prefix" + s[2..4];
```
But these methods pollute the namespace because they show up on any expression (it's the same as extending `object`).

**Comment by RikkiGibson** [comment, marked as answer]
#5495
